### DISCUSSION TITLE
Need more clarity regarding new application

### DISCUSSION THREAD

**Discussion by BkPankaj** [JdeRobot/VisualCircuit#273] [General]
Hello @jmplaza sir,

In reference to your [comment](https://github.com/JdeRobot/VisualCircuit/issues/221#issuecomment-1917645577) on the issue, I have a question regarding the new application mentioned. I'm interested to know are you referring to a Desktop application for VisualCircuit, similar to the [RoboticsAcademy Desktop application](https://github.com/JdeRobot/RoboticsAcademy-Desktop) developed by [Prakhar's Bansal](https://jderobot.github.io/activities/gsoc/) during GSOC 2023.

Could you kindly confirm if my understanding is correct?

**Comment by jmplaza** [comment]
Hi @BkPankaj , 

no, just a robotics application in Gazebo simulator, a piece of software connecting to robot sensor drivers and actuator drivers. Not an electron-based app.

Take a look at previous examples: [computer vision application](https://www.youtube.com/watch?v=0peIsg6ynE8) ,  [FollowLine application](https://www.youtube.com/watch?v=w8S_1Qn6ETI), [another computer vision application](https://www.youtube.com/watch?v=lpSiZNpHGcc), [vacuum cleaner application](https://www.youtube.com/watch?v=zPdjZZFtMjs), [FollowPerson application](https://www.youtube.com/watch?v=IknpvAs_jAo), and [drone application ](https://www.youtube.com/watch?v=J1JhnNOKe1o)

**Comment by BkPankaj** [comment]
Hello @jmplaza,
Thank you for your reply. It is not related to VC+ section, right? I am still unable to understand software that connects to robot sensor and actuator drivers. In the [FollowPerson application](https://www.youtube.com/watch?v=IknpvAs_jAo), it seems like the person is providing actuator commands for position and angular velocity values through a PID block. Are you suggesting the development of software similar to RViz?

**Reply by jmplaza** [reply]
Nope. 

Just a VisualCircuit project, an application that is connected to a simulated robot through some of the existing driver blocks (a driver for motors, a driver for camera images...). I suggest you start replicating a simple robotics application. Maybe a "hello world" one just using the motor block and commanding it to move straight for 2 seconds, and then turn for 3  seconds and then move straight again for other 2 seconds. 
The motor blocks accepts speed commands AFAIK

You have to launch a Gazebo world with the scenario and the simulated robot providing the corresponding ROS topics for motors. Then you may run the project generated from VisualCircuit. The driver blocks will connect to the corresponding topics in case everything is properly configured. And the robotics applicatiion (designed as a collection of blocks) will control the robot motion.

**Reply by BkPankaj** [reply]
Thank you for explaining @jmplaza . I didn't use the Gazebo world, but I will learn and try the hello world of the motor block with the Gazebo world. I will reply back to you after running the simple robotics application with the Gazebo world.

**Reply by BkPankaj** [reply]
Hello @jmplaza ,
I successfully ran a basic robotic application with Gazebo and VisualCircuit, as you suggested. What I observed is that the motor block publishes data to a particular topic. So, I assume you are suggesting to create an application similar to what we do with echo or rqt graph in the terminal, providing information about topics, bandwidth, etc., but specifically for the topics used by VisualCircuit blocks. Is my understanding correct?

I implemented a 2-second run in the positive x-direction for a differential car and a 3-second run in the negative x-direction. Below are the images:

VisualCIrcuit Canvas with required blocks
![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/2fcf3f26-e68f-4126-946b-40b31748396d)

Output in terminal
![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/f9dfd4bd-cecc-4b40-a9e4-ef3d9fbba182)

Inside gazebo at particular time
![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/efa17847-c67e-4913-a0d8-480ab6a5d096)

Inside gazebo after some time
![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/ba3816dc-a5ea-4ffa-aa51-06ebbd514598)


--------------------------------
--------------------------------


I also noticed an error: (velocities != None) cannot be used when dealing with a NumPy array for comparing with None. Instead, we should use (velocities is not None). While this issue may seem small, new users to NumPy might face difficulties resolving it. Therefore, it should be corrected within the MotorDriverROS2 block. I would like to work on addressing this issue. Could you please verify? Below is an image of the error.
![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/2f69b4c0-33ff-4ae1-9709-b90d66d70437)

**Reply by BkPankaj** [reply]
Hi @jmplaza,
Waiting for any reply to this [comment](https://github.com/JdeRobot/VisualCircuit/discussions/273#discussioncomment-8362505). Apologies for any inconvenience caused.

**Reply by jmplaza** [reply]
Uhm... you use VisualCircuit to program the robotics application, just as a combination of blocks. Some of the blocks are already available at the VisualCircuit library (for instance driver blocks or sensor blocks). The user may also define his/her own blocks at will, writing their Python implementation.

In order to run it successfully you first have to launch locally the Gazebo world with the simulated scenario and the robot model, which should include the robot sensor plugins and robot motor plugins. The plugins provide a set of standardized ROS topics. This is the ROS interface of the robot body, its Hardware Abstraction Layer.

The robotics application is the robot brain, which connects to the corresponding ROS topics of the ROS interface. Yes, the existing motor block connects to the corresponding standard ROS topics. The existing camera block connects to the correspoding ROS topic for images. 
 
@siddheshsingh26 @toshan-luktuke do you have any example application?
